pkg/config/template.go

```diff
@@ -116,47 +116,84 @@ func (g osFileGetter) GetFile(filename string) ([]byte, error) {
-				changes = append(changes, prowgithub.PullRequestChange{
-					Filename: path,
-					Status:   string(prowgithub.PullRequestFileModified),
-				})
-			}
+func (c *TemplateCMManager) createCM(dir, name string, data map[string]string) error {
+	cm := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			Labels: map[string]string{
+				createByRehearse:  "true",
+				rehearseLabelPull: strconv.Itoa(c.prNumber),
+			},
+		},
+		Data: map[string]string{},
+	}
+	if _, err := c.cmclient.Create(cm); err != nil && !kerrors.IsAlreadyExists(err) {
+		return err
+	} else if err := updateconfig.Update(osFileGetter{root: dir}, c.cmclient, cm.Name, cm.Namespace, data, c.logger); err != nil {
+		return err
+	}
+	return nil
+}
+
+func genChanges(root, dir string) ([]prowgithub.PullRequestChange, error) {
+	var ret []prowgithub.PullRequestChange
+	err := filepath.Walk(filepath.Join(root, dir), func(path string, info os.FileInfo, err error) error {
+		if err != nil || info.IsDir() {
+			return err
+		}
+		path, err = filepath.Rel(root, path)
+		if err != nil {
 			return err
+		}
+		ret = append(ret, prowgithub.PullRequestChange{
+			Filename: path,
+			Status:   string(prowgithub.PullRequestFileModified),
 		})
+		return nil
+	})
+	return ret, err
+}
+
+func replaceSpecNames(spec map[string]prowplugins.ConfigMapSpec, mapping map[string]string) map[string]prowplugins.ConfigMapSpec {
+	ret := make(map[string]prowplugins.ConfigMapSpec, len(spec))
+	for k, v := range spec {
+		if name, ok := mapping[v.Name]; ok {
+			v.Name = name
+		}
+		ret[k] = v
+	}
+	return ret
+}
+
+func (c *TemplateCMManager) createCMs(dir string, filenames []string, cmSpec map[string]prowplugins.ConfigMapSpec) error {
+	var changes []prowgithub.PullRequestChange
+	for _, f := range filenames {
+		c, err := genChanges(dir, f)
 		if err != nil {
 			return err
 		}
+		changes = append(changes, c...)
 	}
 	var errs []error
-	for cm, data := range updateconfig.FilterChanges(cms, changes, c.logger) {
-		profile := strings.TrimPrefix(cm.Name, "cluster-profile-")
-		for _, p := range profiles {
-			if p.Name == profile {
-				profile = p.CMName()
-				break
-			}
-		}
-		err := c.createCM(&v1.ConfigMap{
-			ObjectMeta: metav1.ObjectMeta{Name: profile},
-			Data:       map[string]string{},
-		})
-		if err != nil && !kerrors.IsAlreadyExists(err) {
-			errs = append(errs, err)
-		} else if err := updateconfig.Update(osFileGetter{root: dir}, c.cmclient, profile, cm.Namespace, data, c.logger); err != nil {
+	for cm, data := range updateconfig.FilterChanges(cmSpec, changes, c.logger) {
+		c.logger.WithFields(logrus.Fields{"cm-name": cm.Name}).Info("creating rehearsal configMap")
+		if err := c.createCM(dir, cm.Name, data); err != nil {
 			errs = append(errs, err)
 		}
 	}
 	return kutilerrors.NewAggregate(errs)
 }
 
+func (c *TemplateCMManager) CreateClusterProfiles(dir string, cmSpec map[string]prowplugins.ConfigMapSpec, profiles []ClusterProfile) error {
```

Comment:
I haven't noticed this in #127, but I would slightly prefer to have the path/config items in the manager struct to keep Create* method API dead simple. I think the struct can keep the "static stuff that we need to create templates/CMs" and Create* methods can just take params what to create.

Reply:
Good call. Both PRs updated.